# Data Management and Analysis (CSCI-UA 479)
## Erin Choi - Assignment 1, Part 2: Books

### Books used:
#### *The Great Gatsby* by F. Scott Fitzgerald
https://www.gutenberg.org/ebooks/64317

**Licensing information:**

This eBook is for the use of anyone anywhere in the United States and
most other parts of the world at no cost and with almost no restrictions
whatsoever. You may copy it, give it away or re-use it under the terms
of the Project Gutenberg License included with this eBook or online at
www.gutenberg.org. If you are not located in the United States, you
will have to check the laws of the country where you are located before
using this eBook.

**Encoding**: UTF-8

#### *Little Women* by Louisa May Alcott
https://www.gutenberg.org/ebooks/514

**Licensing information:**

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Project Gutenberg License included
with this eBook or online at www.gutenberg.org

**Encoding:** UTF-8

### Questions about books:
- *The Great Gatsby*: Which character, besides Jay Gatsby, is mentioned by name most?
    * Characters included in my analysis (only using mentions of first names) are Nick (Carraway, narrator), Daisy (Buchanan), Tom (Buchanan), Jordan (Baker), Myrtle (Wilson), George (Wilson), and Meyer (Wolfsheim).
- *Little Women*: How many times are each of the 4 sisters mentioned by name?
    * The sisters' names are Meg, Jo, Beth, and Amy.

**Note about reading books:**
Since the encoding for both books is UTF-8, the books can be read in as usual without changing the encoding parameter of the file open function.

### Analysis of *The Great Gatsby*

First, I read in the lines of the book and store the lines in a list.

In [37]:
# read lines in book, store as a list
gatsby = open('data/gatsby.txt', 'r')
gatsby_lines = gatsby.readlines()
gatsby.close()

For each line in the list of lines:
* I remove the white space around the lines to get rid of new lines, tabs, etc.
* I split the line at each space to get a list of words in that line.
* I remove the starting and end punctuation by replacing it with an empty string. (I don't examine each character in a "word" to see if it's punctuation, since that could affect contractions, compound words connected by dashes, etc.)
* I append each non-empty "word" to a list of words in the whole book.

In [38]:
# remove white space, separate words, remove punctuation

gatsby_lines = list(map(str.strip, gatsby_lines))
gatsby_words = []
punctuation = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''

for line in gatsby_lines:
    
    # split each line at spaces to get words
    line_words = line.split()
    for word in line_words:
        # replace only starting and end punctuation with empty string 
        if word[-1] in punctuation:
            word = word.replace(word[-1], '')
        if word!='' and word[0] in punctuation:
            word = word.replace(word[0], '')
        
        # append non-empty "words" to list of words
        if word!='':
            gatsby_words.append(word)

Finally, I count the number of occurrences of each name of interest by looping through the list of all words and update a dictionary, with keys being the names and values being the number of the name's occurrences.

In [39]:
# count the number of occurrences of each name of interest
# by looping through all words and updating a dictionary
name_dict = {}
name_list = ['Nick', 'Daisy', 'Tom', 'Jordan', 'Myrtle', 'Meyer']

for word in gatsby_words:
    if word in name_list:
        if word in name_dict.keys():
            name_dict[word] += 1
        else:
            name_dict[word] = 1

name_dict

{'Tom': 164, 'Daisy': 137, 'Nick': 13, 'Jordan': 61, 'Myrtle': 18, 'Meyer': 5}

**The results of my analysis show that Tom Buchanan's name is mentioned the most - he is named 164 times in *The Great Gatsby*.** I thought Daisy would be mentioned the most besides Gatsby since she is Gatsby's love interest, but surprisingly, Tom is mentioned more. (Or maybe not surprisingly, since he is Daisy's husband and thus is in a position of conflict with Gatsby...)

### Analysis of *Little Women*

Again, I read in the lines of the book and store them in a list.

In [40]:
lw = open('data/littlewomen.txt', 'r')
lw_lines = lw.readlines()
lw.close()

I repeat the steps I performed for *The Great Gatsby* here.
For each line, I strip white space, obtain a list of words in the line, remove starting and end punctuation, and append non-empty words to a list.

In [42]:
lw_lines = list(map(str.strip, lw_lines))
lw_words = []

for line in lw_lines:

    line_words = line.split()
    for word in line_words:
        if word[-1] in punctuation:
            word = word.replace(word[-1], '')
        if word!='' and word[0] in punctuation:
            word = word.replace(word[0], '')
        
        if word!='':
            lw_words.append(word)

I count the number of times each sister ("Little Woman") is mentioned by looping through the list of all words and update a dictionary.

In [43]:
lw_name_dict = {}
woman_list = ['Meg', 'Jo', 'Beth', 'Amy']

for word in lw_words:
    if word in woman_list:
        if word in lw_name_dict.keys():
            lw_name_dict[word] += 1
        else:
            lw_name_dict[word] = 1

lw_name_dict

{'Jo': 1209, 'Meg': 615, 'Amy': 553, 'Beth': 393}

The results are as follows:
* **Meg is mentioned 615 times.**
* **Jo is mentioned 1209 times.**
* **Beth is mentioned 393 times.**
* **Amy is mentioned 553 times.**

This makes sense because although the story is about all of the sisters, Jo is in a way the "main" sister that the novel focuses on and follows most closely, particularly towards the end of the book. **Spoiler Alert:** Beth is likely mentioned the least because she was not as active as her sisters, and she passes away partway through the book.